File: src/engraft/engine.py
import shutil
from pathlib import Path

from engraft.models import parse_template, parse_values

STAGING_DIR_NAME = ".engraft"


def apply(
    template_path: str | Path,
    values_path: str | Path,
    work_dir: Path | None = None,
) -> None:
    """Apply a template with the given values to the working directory.

    Uses a staging directory (.engraft/) to ensure atomicity: all actions
    are executed against the staging copy first, and results are only
    copied back to the project if all actions succeed.
    """
    template_path = Path(template_path).resolve()
    values_path = Path(values_path).resolve()

    if work_dir is None:
        work_dir = Path.cwd()
    else:
        work_dir = Path(work_dir).resolve()

    values_dir = values_path.parent

    template = parse_template(template_path)
    values = parse_values(values_path)

    # Resolve variables: values override defaults
    variables: dict[str, str] = {}
    for name, var in template.variables.items():
        variables[name] = values.get(name, var.default)

    staging_dir = work_dir / STAGING_DIR_NAME

    # Step 1: Collect target files from all actions
    target_files: set[str] = set()
    for action in template.customizations:
        target_files.update(action.target_files())

    # Step 2: Create staging directory (remove if exists)
    if staging_dir.exists():
        shutil.rmtree(staging_dir)
    staging_dir.mkdir()

    try:
        # Copy target files into staging directory
        for rel_path in target_files:
            src = work_dir / rel_path
            dst = staging_dir / rel_path
            if src.exists():
                dst.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy2(src, dst)

        # Step 3: Execute all actions against the staging directory
        for action in template.customizations:
            action.apply(variables, staging_dir, values_dir)

        # Step 4: Copy back all files from staging to work directory
        for staged_file in staging_dir.rglob("*"):
            if staged_file.is_file():
                rel_path = staged_file.relative_to(staging_dir)
                dest = work_dir / rel_path
                dest.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy2(staged_file, dest)
    finally:
        # Step 5: Always remove staging directory
        shutil.rmtree(staging_dir, ignore_errors=True)

File: src/engine.ts
import { promises as fs, existsSync } from "node:fs";
import path from "node:path";

import { parseTemplate, parseValues } from "./models.js";
import type { Variables } from "./actions/base.js";

export const STAGING_DIR_NAME = ".engraft";

interface ApplyOptions {
  workDir?: string;
}

const rmrf = async (dir: string): Promise<void> => {
  await fs.rm(dir, { recursive: true, force: true });
};

const copyFile = async (src: string, dst: string): Promise<void> => {
  await fs.mkdir(path.dirname(dst), { recursive: true });
  await fs.copyFile(src, dst);
};

const walkFiles = async (dir: string): Promise<string[]> => {
  const out: string[] = [];
  const entries = await fs.readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      out.push(...(await walkFiles(full)));
    } else if (entry.isFile()) {
      out.push(full);
    }
  }
  return out;
};

export const apply = async (
  templatePath: string,
  valuesPath: string,
  options: ApplyOptions = {},
): Promise<void> => {
  const absTemplate = path.resolve(templatePath);
  const absValues = path.resolve(valuesPath);
  const workDir = path.resolve(options.workDir ?? process.cwd());
  const valuesDir = path.dirname(absValues);

  const template = parseTemplate(absTemplate);
  const values = parseValues(absValues);

  const variables: Variables = {};
  for (const [name, variable] of Object.entries(template.variables)) {
    if (Object.prototype.hasOwnProperty.call(values, name)) {
      variables[name] = values[name] ?? null;
    } else {
      variables[name] = variable.default;
    }
  }

  const stagingDir = path.join(workDir, STAGING_DIR_NAME);

  const targetFiles = new Set<string>();
  for (const action of template.customizations) {
    for (const rel of action.targetFiles()) {
      targetFiles.add(rel);
    }
  }

  if (existsSync(stagingDir)) {
    await rmrf(stagingDir);
  }
  await fs.mkdir(stagingDir);

  try {
    for (const rel of targetFiles) {
      const src = path.join(workDir, rel);
      const dst = path.join(stagingDir, rel);
      if (existsSync(src)) {
        await copyFile(src, dst);
      }
    }

    for (const action of template.customizations) {
      await action.apply(variables, stagingDir, valuesDir);
    }

    const stagedFiles = await walkFiles(stagingDir);
    for (const staged of stagedFiles) {
      const rel = path.relative(stagingDir, staged);
      const dest = path.join(workDir, rel);
      await copyFile(staged, dest);
    }
  } finally {
    await rmrf(stagingDir);
  }
};
